src/lib/pubkey/sphincsplus/sphincsplus_common/sphincsplus.cpp

```diff
@@ -0,0 +1,340 @@
+            m_public(std::move(pub_key)),
+            m_hashes(Botan::Sphincs_Hash_Functions::create(m_public->parameters(), m_public->seed())) {}
+
+      /**
+       * Add more data to the message currently being signed
+       * @param msg the message
+       * @param msg_len the length of msg in bytes
+       */
+      void update(const uint8_t msg[], size_t msg_len) override {
+         m_msg_buffer.insert(m_msg_buffer.end(), msg, msg + msg_len);
+      }
+
+      /*
+      * Perform a verification operation
+      * @param rng a random number generator
+      */
+      bool is_valid_signature(const uint8_t* sig, size_t sig_len) override {
+         const auto& p = m_public->parameters();
+         if(sig_len != p.sphincs_signature_bytes()) {
+            return false;
+         }
+
+         BufferSlicer s({sig, sig_len});
+         // Compute leaf and tree index from R
+         const auto msg_random_s = s.take<SphincsMessageRandomness>(p.n());
+         auto [mhash, tree_idx, leaf_idx] = m_hashes->H_msg(msg_random_s, m_public->root(), m_msg_buffer);
+         // Clear the message buffer, the data is not needed anymore
+         m_msg_buffer.clear();
+
+         // Reconstruct the FORS tree
+         Sphincs_Address fors_addr(Sphincs_Address_Type::ForsTree);
+         fors_addr.set_tree(tree_idx).set_keypair(leaf_idx);
+         const auto fors_sig_s = s.take<ForsSignature>(p.fors_signature_bytes());
+         auto fors_root = fors_public_key_from_signature(mhash, fors_sig_s, fors_addr, p, *m_hashes);
+
+         // Verify the hypertree signature
+         const auto ht_sig_s = s.take<SphincsHypertreeSignature>(p.ht_signature_bytes());
+         BOTAN_ASSERT_NOMSG(s.empty());
+         return ht_verify(fors_root, ht_sig_s, m_public->root(), tree_idx, leaf_idx, p, *m_hashes);
+      }
+
+      std::string hash_function() const override { return m_hashes->msg_hash_function_name(); }
+
+   private:
+      std::shared_ptr<SphincsPlus_PublicKeyInternal> m_public;
+      std::unique_ptr<Sphincs_Hash_Functions> m_hashes;
+      std::vector<uint8_t> m_msg_buffer;
+};
+
+std::unique_ptr<PK_Ops::Verification> SphincsPlus_PublicKey::create_verification_op(std::string_view /*params*/,
+                                                                                    std::string_view provider) const {
+   if(provider.empty() || provider == "base")
+      return std::make_unique<SphincsPlus_Verification_Operation>(m_public);
+   throw Provider_Not_Found(algo_name(), provider);
+}
+
+std::unique_ptr<PK_Ops::Verification> SphincsPlus_PublicKey::create_x509_verification_op(
+   const AlgorithmIdentifier& signature_algorithm, std::string_view provider) const {
+   if(provider.empty() || provider == "base") {
+      if(signature_algorithm != this->algorithm_identifier())
+         throw Decoding_Error("Unexpected AlgorithmIdentifier for SPHINCS+ signature");
```

Comment:
This is missing braces, which will make `clang-tidy` unhappy. This won't be caught by CI until after #3571 is merged

Reply:
In the meantime I'd suggest doing a full `clang-tidy` run and addressing any issues.

Reply:
Done!